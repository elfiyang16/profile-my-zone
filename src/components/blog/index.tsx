import React, { FunctionComponent } from "react"
import { RichText } from "../richText"
import { IBlog } from "../../types/IBlog"
import moment from "moment"
import "./styles.scss"

const Blog: FunctionComponent<IBlog> = data => {
  const { title, body, publishDate, heroImage } = data

  return (
    <div className="article">
      <div className="article__img-container">
        <img
          className="article__img"
          src={heroImage.file.url}
          alt={heroImage.title}
        />
        <small className="article__img-title">{heroImage.title}</small>
      </div>
      <article className="article__content">
        <h2 className="article__content-title">{title}</h2>
        <p className="article__content-date">
          {moment(publishDate).format("LL")}
        </p>
        <RichText content={body} />
      </article>
    </div>
  )
}

export default Blog
