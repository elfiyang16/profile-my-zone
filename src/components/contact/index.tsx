import React, { useState } from "react"
import "./styles.scss"

enum State {
  SUCCESS = "SUCCESS",
  ERROR = "ERROR",
}

const Contact: React.FunctionComponent = () => {
  const [status, setStatus] = useState<State | undefined>(undefined)

  const submitForm = (ev: any) => {
    ev.preventDefault()
    const form = ev.target
    const data = new FormData(form)
    const xhr = new XMLHttpRequest()
    xhr.open(form.method, form.action)
    xhr.setRequestHeader("Accept", "application/json")
    xhr.onreadystatechange = () => {
      if (xhr.readyState !== XMLHttpRequest.DONE) return
      if (xhr.status === 200) {
        form.reset()
        setStatus(State.SUCCESS)
      } else {
        setStatus(State.ERROR)
      }
    }
    xhr.send(data)
  }

  const returnAction = (status: State) => {
    if (status === "SUCCESS") {
      return (
        <p className="contact__submit-result">
          Thanks for contacting me. I will come back to you soon!
        </p>
      )
    }
    if (status === "ERROR") {
      return (
        <p className="contact__submit-result">Ooops! There was an error.</p>
      )
    }
    return (
      <button className="contact__submit" type="submit">
        Submit
      </button>
    )
  }

  return (
    <form
      className="contact__form"
      onSubmit={submitForm}
      action={`${process.env.GATSBY_FORM_ENDPOINT}`}
      method="POST"
    >
      <label className="contact__label">Name:</label>
      <input className="contact__input" type="name" name="name" />
      <label className="contact__label">Email:</label>
      <input className="contact__input" type="email" name="email" />
      <label className="contact__label">Message:</label>
      <textarea className="contact__text" name="message" />
      {returnAction(status)}
    </form>
  )
}

export default Contact
