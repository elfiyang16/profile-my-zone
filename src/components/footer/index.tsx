import React from "react"
import "./styles.scss"

const Footer: React.FunctionComponent = () => {
  return (
    <footer className="site-footer">
      Â© {new Date().getFullYear()}, built by Elfi with Gatsby under{" "}
      <a href="https://opensource.org/licenses/MIT">MIT License</a>
    </footer>
  )
}

export default Footer
