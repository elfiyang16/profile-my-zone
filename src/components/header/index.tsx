import { Link, navigate } from "gatsby"
import React, { useState } from "react"
import "./styles.scss"
import dog from "../../images/dog-white.svg"
import burgermenu from "../../images/menu-white.svg"
import closeicon from "../../images/close-icon.png"
import { set } from "lodash"

interface IHeader {
  header: string
}

interface IMenu {
  url: string
  item: string
  menu: boolean
  setMenu: React.Dispatch<React.SetStateAction<boolean>>
}

const CustomLink: React.FC<IMenu> = ({ menu, setMenu, url, item }) => (
  <Link
    to={`${url}`}
    onClick={(e: any) => {
      e.preventDefault()
      setMenu(!menu)
      navigate(url)
    }}
    className="site-header__menu-item"
  >
    {item}
  </Link>
)

const Header: React.FunctionComponent<IHeader> = ({ header }) => {
  const [menu, setMenu] = useState<boolean>(false)

  return (
    <header className={`site-header ${menu ? "active" : ""}`}>
      <div className="site-header__container">
        <div className="site-header__main">
          <div>
            <Link to="/">
              <img src={dog} alt="home logo" className="site-header__logo" />
            </Link>
          </div>
          <nav
            className="site-header__responsive-menu-bar"
            tabIndex={0}
            onClick={(e: any) => setMenu(!menu)}
          >
            <button>
              <img
                src={burgermenu}
                alt="menu logo"
                className="site-header__burger-menu"
              />
            </button>
            {header === "home" ? (
              <div className="site-header__menu">
                <button className="site-header__menu-close">
                  <img src={closeicon} alt="close" />
                </button>
                <CustomLink setMenu={setMenu} menu={menu} url="/" item="Home" />
                <CustomLink
                  setMenu={setMenu}
                  menu={menu}
                  url="/#blogs"
                  item="Blogs"
                />
                <CustomLink
                  setMenu={setMenu}
                  menu={menu}
                  url="/#about"
                  item="About me"
                />
                <CustomLink
                  setMenu={setMenu}
                  menu={menu}
                  url="/#contact"
                  item="Contact"
                />
              </div>
            ) : (
              <div className="site-header__menu">
                <CustomLink
                  setMenu={setMenu}
                  menu={menu}
                  url="/#home"
                  item="Home"
                />
              </div>
            )}
          </nav>
        </div>
      </div>
    </header>
  )
}

export default Header
