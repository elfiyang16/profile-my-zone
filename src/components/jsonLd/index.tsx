import React, { FunctionComponent } from "react"
import { Thing, WithContext, BlogPosting } from "schema-dts"
import { IBlog } from "../../types/IBlog"

export interface IJsonLD<T extends Thing> {
  data: WithContext<T>
}
export class JsonLd<T extends Thing> extends React.Component<IJsonLD<T>> {
  public render() {
    const { data } = this.props

    return (
      <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{ __html: JSON.stringify(data) }}
      />
    )
  }
}

export function createBlogJsonLd(post: IBlog): BlogPosting {
  return {
    "@type": "BlogPosting",
    author: "Elfi",
    datePublished: post.publishDate,
    description: post.description ? post.description.description : undefined,
    headline: post.title,
    image: post.heroImage ? post.heroImage.file.url : undefined,
    // mainEntityOfPage: `https://localhost:8000/blog/${post.slug}`,
  }
}

export interface IStructuredData {
  post: IBlog
}

export const StructuredData: FunctionComponent<IStructuredData> = ({
  post,
}) => (
  <JsonLd<BlogPosting>
    data={{
      "@context": "https://schema.org",
      ...createBlogJsonLd(post),
    }}
  />
)
