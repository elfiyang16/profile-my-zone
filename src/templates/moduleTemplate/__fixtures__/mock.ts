import { IHero } from "../../../types/IHero"
import { IBlog } from "../../../types/IBlog"

export const heroMock: any = {
  __typename: "ContentfulHero",
  id: "abc",
  colorTheme: "",
  heading1: "Some title",
  heading2: "Some other title",
  image: {
    title: "The example",
    file: {
      url: "//images.ctfassets.net/example.png",
    },
  },
}

export const gqlResponse: any = {
  contentfulModuleTemplate: {
    id: "abdfd",
    pageTitle: "Testing",
    urlPath: "/test-new-modular-template",
    hero: {
      id: "abc",
      colorTheme: "",
      heading1: "Some title",
      heading2: "Some other title",
      image: {
        title: "The example",
        file: {
          url: "//images.ctfassets.net/example.png",
        },
      },
    },
    modules: [
      {
        __typename: "ContentfulBlogPost",
        id: "bhdkf",
        publishDate: "2020-11-22T00:00+00:00",
        heroImage: {
          title: "Photo on Unsplash",
          file: {
            url: "//images.ctfassets.net/abe.jpeg",
          },
        },
        body: {
          raw: '{"nodeType":"document","data":{},"content":[]}',
        },
        description: {
          description: "description",
        },
        author: { name: "elfi" },
        slug: "/abv",
      },
    ],
  },
}
