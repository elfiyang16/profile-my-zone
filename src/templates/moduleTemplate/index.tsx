import { graphql } from "gatsby"

import ModuleTemplate from "./moduleTemplate"

export const pageQuery = graphql`
  query($id: String!) {
    contentfulModuleTemplate(id: { eq: $id }) {
      id
      pageTitle
      urlPath
      structuredData {
        internal {
          type
          content
        }
      }
      noIndex: doNotIndex
      hero {
        ...hero
      }
      blogs {
        __typename
        ... on Node {
          ...blog
        }
      }
    }
  }

  fragment hero on ContentfulHero {
    id
    colorTheme
    heading1
    heading2
    image {
      title
      file {
        contentType
        fileName
        url
      }
    }
    entryTitle
  }

  fragment blog on ContentfulBlogPost {
    id
    author {
      name
    }
    body {
      raw
    }
    publishDate
    slug
    title
    description {
      description
    }
    heroImage {
      title
      file {
        contentType
        fileName
        url
      }
    }
  }
`

export default ModuleTemplate
