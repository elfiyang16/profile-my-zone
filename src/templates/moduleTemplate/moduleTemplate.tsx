/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { FunctionComponent } from "react"
import Hero from "../../components/hero"
import Blog from "../../components/blog"
import SEO from "../../components/seo"
import Layout from "../../components/layout"
import { IBlog } from "../../types/IBlog"
import { IHero } from "../../types/IHero"
import { JsonLd } from "../../components/jsonLd"

// import "./styles.scss"

type IComponent = IBlog /* | increment with any other modules */

/* | increment with any other modules */
/* right now IModule === IBlogs */
/* right now IComponent === IBlog */
type IModule<V extends IComponent> = {
  __typename: string
  id: string
} & V

interface IStructuredData {
  internal: {
    content: string
  }
}

interface IProps {
  data: {
    contentfulModuleTemplate: {
      pageTitle?: string
      urlPath?: string
      structuredData?: IStructuredData
      metaDescription?: { metaDescription: string }
      noIndex?: boolean
      hero?: IHero
      blogs?: IModule<IComponent>[]
    }
  }
}

export const componentByTypeName = (typename: string, props: IBlog) => {
  const modules: any = {
    /* eslint-disable-next-line react/display-name  */
    ContentfulBlogPost: ({
      id,
      author,
      slug,
      title,
      description,
      heroImage,
      body,
      publishDate,
    }: any) => (
      <Blog
        key={id}
        author={author}
        slug={slug}
        title={title}
        description={description}
        heroImage={heroImage}
        body={body}
        publishDate={publishDate}
      />
    ),
  }
  const Component = modules[typename]
  return Component ? Component(props) : null
}

export const Main: FunctionComponent<IProps> = ({ data }) => {
  const {
    hero,
    blogs,
    structuredData,
    // pageTitle,
    // urlPath,
    // noIndex,
  } = data.contentfulModuleTemplate

  return (
    <div className="modular-content_container">
      {structuredData &&
        structuredData.internal &&
        structuredData.internal.content && (
          <JsonLd data={JSON.parse(structuredData.internal.content)} />
        )}
      {hero && <Hero {...hero} />}
      {Array.isArray(blogs) &&
        blogs.map(m => {
          const { __typename, ...props } = m
          return componentByTypeName(__typename, props)
        })}
    </div>
  )
}

const ModuleTemplate: FunctionComponent<IProps> = ({ data }) => {
  const { urlPath, pageTitle, noIndex } = data.contentfulModuleTemplate

  return (
    <Layout>
      <SEO title={pageTitle} path={urlPath} noIndex={!!noIndex} />
      <Main data={data} />
    </Layout>
  )
}

export default ModuleTemplate
