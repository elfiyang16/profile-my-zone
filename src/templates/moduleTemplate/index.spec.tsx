import { shallow } from "enzyme"
import { StaticQuery } from "gatsby"
import { cloneDeep } from "lodash"
import React from "react"
import { JsonLd } from "../../components/jsonLd"
import { Main } from "./moduleTemplate"
import { gqlResponse } from "./__fixtures__/mock"

jest.mock("gatsby")

let data: any
beforeEach(() => {
  data = cloneDeep(gqlResponse)
  ;(StaticQuery as jest.Mock).mockImplementationOnce(({ render }) =>
    render({
      site: {
        siteMetadata: {
          title: "elfi dev zone",
        },
      },
    })
  )
})

describe("Main", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Main data={data as any} />)

    expect(wrapper).toMatchSnapshot()
  })

  it("should render <JsonLd /> correctly if structuredData is passed", () => {
    const newData = {
      contentfulModuleTemplate: {
        ...data.contentfulModuleTemplate,
        structuredData: {
          internal: {
            content: `{
              "@type": "Blog",
              "@context": "https://schema.org",
              "itemListElement": [
                {
                  "item": { "@id": "https://elfiy.com", "name": "Home" },
                  "@type": "ListItem",
                  "position": 1
                }
              ]
            }`,
          },
        },
      },
    }
    const wrapper = shallow(<Main data={newData as any} />)

    expect(wrapper.find(JsonLd).length).toBe(1)
  })
})
