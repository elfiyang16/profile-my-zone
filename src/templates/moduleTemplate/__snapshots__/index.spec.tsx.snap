// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
<div
  className="modular-content_container"
>
  <Hero
    colorTheme=""
    heading1="Some title"
    heading2="Some other title"
    id="abc"
    image={
      Object {
        "file": Object {
          "url": "//images.ctfassets.net/example.png",
        },
        "title": "The example",
      }
    }
  />
</div>
`;
